PR: Fix hover/calltip for Python objects without signature

goanpeca opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes


* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

![hints](https://user-images.githubusercontent.com/3627835/57564411-b9aa7c00-7370-11e9-8a91-49a53bc81d35.gif)


<!--- Explain what you've done and why --->




### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #9294

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @goanpeca 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @goanpeca! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`spyder/widgets/mixins.py`](https://github.com/spyder-ide/spyder/blob/e9b9c2677508b271f4e173a7a821e0c393e189d1/spyder/widgets/mixins.py):

> [Line 341:32](https://github.com/spyder-ide/spyder/blob/e9b9c2677508b271f4e173a7a821e0c393e189d1/spyder/widgets/mixins.py#L341): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 371:50](https://github.com/spyder-ide/spyder/blob/e9b9c2677508b271f4e173a7a821e0c393e189d1/spyder/widgets/mixins.py#L371): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 377:1](https://github.com/spyder-ide/spyder/blob/e9b9c2677508b271f4e173a7a821e0c393e189d1/spyder/widgets/mixins.py#L377): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

goanpeca commented:
@ccordoba12 could you check if you find any other issues?

ccordoba12 commented:
> could you check if you find any other issues?

I think it's time to add some tests for hovers. Perhaps by calling `request_hover` of CodeEditor, then waiting for `sig_display_signature` to be emitted and passing its arg to `show_tooltip`?

goanpeca commented:
>I think it's time to add some tests for hovers. Perhaps by calling request_hover of CodeEditor, then waiting for sig_display_signature to be emitted and passing its arg to show_tooltip

Sure, I will try to add checks for these methods

ccordoba12 commented:
@goanpeca, let's leave tests for later. I have one important remark though: the hover and tooltip for `dict` is too wide. Could you wrap its text with `textwrap` so all hovers/calltips have the same width?

goanpeca commented:
Yes, I noticed I will make the changes

ccordoba12 commented:
Great, thanks!

goanpeca commented:
![hints](https://user-images.githubusercontent.com/3627835/57564411-b9aa7c00-7370-11e9-8a91-49a53bc81d35.gif)

Fixed issues and added tests @ccordoba12 (I marked them as slow and second, but they are actually not that slow)

goanpeca commented:
I noticed the calltip test is currently being performed on the main window, so maybe we could remove that one (when these ones are more complete, as they will run faster on the code editor fixture alone)

goanpeca commented:
Trying to run these tests on my Kubuntu vm is yielding:

```python

```

goanpeca commented:
Yeiii checks passed. So mouse move events are too flaky on CI. They worked locally though 🤷🏻‍♂️

ccordoba12 commented on spyder/plugins/editor/widgets/codeeditor.py:
```diff
@@ -275,11 +275,11 @@ class CodeEditor(TextEditBaseWidget):
 
     #: Signal to display object information on the Help plugin
-    sig_display_object_info = Signal(str)
+    sig_display_object_info = Signal(object)
```
I don't understand this change. Why is it needed?

goanpeca replied:
I was testing something out, but my question is: The LSP will always return `str` on python 2.7 ?

ccordoba12 replied:
I'm not sure. In that case this change would be good.

ccordoba12 commented on spyder/plugins/editor/widgets/codeeditor.py:
```diff
@@ -275,11 +275,11 @@ class CodeEditor(TextEditBaseWidget):
     #: Signal only used for tests
     # TODO: Remove it!
-    sig_signature_invoked = Signal()
+    sig_signature_invoked = Signal(object)
```
Also this one?

goanpeca replied:
Same as above, but in this case I need to be able to check on the tests what the reply was 

https://github.com/spyder-ide/spyder/pull/9301/files#diff-d6fbacae4782021d26229f59f1a4a082R101

ccordoba12 commented on spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py:
```diff
@@ -0,0 +1,106 @@
+some_function""".format(SIG=TEST_SIG, DOC=TEST_DOCSTRING)
+
+
+@pytest.mark.slow
```
Please also mark this test as `second` (as it's done in `test_warnings`).

goanpeca replied:
Will do

ccordoba12 commented on spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py:
```diff
@@ -0,0 +1,106 @@
+        assert expected_output_text in signatures['label']
+
+
+# @pytest.mark.slow
```
Please revert this and mark as `second`.

goanpeca replied:
Will do

ccordoba12 commented on spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py:
```diff
@@ -0,0 +1,106 @@
+
+    with qtbot.waitSignal(code_editor.sig_display_object_info,
+                          timeout=30000) as blocker:
+        code_editor.hide_tooltip()
```
This should be put before the `with`, right?

goanpeca replied:
This might not be needed anymore now that I am checking the output of the signal and not the visibility of the hover hints

ccordoba12 commented on spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py:
```diff
@@ -0,0 +1,106 @@
+        cursor = code_editor.cursorForPosition(point)
+        line, col = cursor.blockNumber(), cursor.columnNumber()
+        code_editor.request_hover(line, col)
+        qtbot.wait(3000)
```
Is this line really necessary? I mean, immediately after the PyLS response comes to Spyder `sig_display_object_info` is emitted, so there shouldn't be any need fo wait longer here.

goanpeca replied:
Yes, agreed (this is a left over from the checks that where actually moving the mouse)

goanpeca replied:
Actually there is a need for these waits otherwise the content of the emitted signal seems to be none and the tests fail, so I added this back.

ccordoba12 commented on spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py:
```diff
@@ -0,0 +1,106 @@
+        # qtbot.mouseMove(code_editor, pos=point, delay=2000)
+        # code_editor.document_did_change('')
+        # qtbot.mouseClick(code_editor, Qt.LeftButton, pos=point, delay=2000)
+        # assert code_editor.tooltip_widget.isVisible()
```
Please remove these commented lines if they are not necessary.

goanpeca replied:
Will do

ccordoba12 commented:
@goanpeca, sorry for the multiple comments, I forgot to start this as a proper review.

goanpeca commented:
>@goanpeca, sorry for the multiple comments, I forgot to start this as a proper review.


No problem. Will make the fixes now 👍
